KS-15-02-Weather-Station-Stats

DenisGubernatorov opened the pull request:
1) Реализация абстрактного класса WeatherStationStats имплементирующего интерфейс DataSend с методом отправки данных 2) Класс Temperature, переопределяющий логику отправки данных 3) 2) Класс PrecipitationAmount, переопределяющий логику отправки данных

DenisGubernatorov committed:
KS-15-02-Weather-Station-Stats
1) Реализация абстрактного класса WeatherStationStats имплементирующего интерфейс DataSend с методом отправки данных
2) Класс Temperature, переопределяющий логику отправки данных
3) 2) Класс PrecipitationAmount, переопределяющий логику отправки данных

ievetrovdev commented on src/main/kotlin/lesson15/lesson15_task2.kt:
```diff
@@ -1 +1,26 @@
+
+abstract class WeatherStationStats : DataSend
+
+interface DataSend {
```
Зачем нам интерфейс в данной задаче?)

DenisGubernatorov replied:
Интерфейс убрал. В классе-родителе добавил абстрактный метод отправки данных.

ievetrovdev reviewed (changes requested):
Не соответствует ТЗ, давай прочитаем ТЗ еще раз и сделаем как там сказано)

DenisGubernatorov committed:
KS-15-02-Weather-Station-Stats
1) Убран интерфейс отвечающий за реализацию метода отправки данных
2) Метод отправки данных сделан абстрактным в классе-родителе WeatherStationStats. В классах наследниках переопределяется

ievetrovdev commented on src/main/kotlin/lesson15/lesson15_task2.kt:
```diff
@@ -1 +1,24 @@
 package lesson15
+
+abstract class WeatherStationStats {
```
Этот абстрактный класс используется в качестве базового класса, от которого наследуются более конкретные классы, описывающие данные метеостанции. В данной задачи это необходимо для демонстрации полиморфизма.

ievetrovdev commented on src/main/kotlin/lesson15/lesson15_task2.kt:
```diff
@@ -1 +1,24 @@
 package lesson15
+
+abstract class WeatherStationStats {
+    abstract fun doSendData()
```
В ТЗ написано `Реализуй логику имитации отправки сообщения на сервер в классе WeatherServer`
Данные метеостанции не должны отвечать за отправку самих себя)

ievetrovdev commented on src/main/kotlin/lesson15/lesson15_task2.kt:
```diff
@@ -1 +1,24 @@
+    abstract fun doSendData()
+}
+
+class Temperature : WeatherStationStats() {
```
Поскольку данные температуры и количества осадков обычно содержат какие-то значения, предлагаю добавить соответствующие поля в эти классы)

ievetrovdev reviewed (changes requested):
Все же рекомендую перечитать ТЗ еще раз, оставил комментов

DenisGubernatorov committed:
KS-15-02-Weather-Station-Stats
1) Рефакторинг согласно ТЗ
2) Добавлен класс WeatherServer, отвечающий за отправку сообщений.

ievetrovdev commented on src/main/kotlin/lesson15/lesson15_task2.kt:
```diff
@@ -1 +1,45 @@
+
+class WeatherServer {
+    fun sendMessage(message: WeatherStationStats) {
+        when (message) {
```
Хорошо, но предлагаю для большей безопасности учесть вариант, что могут прийти данные, не относящиеся к этим двум классам.

DenisGubernatorov committed:
KS-15-02-Weather-Station-Stats
1) При отправке данных добавлено сообщение для неизвестного типа данных
2) Для тестов добавлен класс Freezing
3) Поправлено сообщения для отправки количества осадков. Вместо объекта PrecipitationAmount выводится id станции с которой они получены

ievetrovdev reviewed (approved):
Теперь корректно, аппрув!
